### PR TITLE
DOC 8527 - Update sample app to support agent scope

### PR DESCRIPTION
**Pull request by maria-robobug**

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    #   - CB_HOST
+    #   - CB_USER
+    #   - CB_PSWD
+    # command: $CB_HOST $CB_USER $CB_PSWD
+    volumes:
+      - .:/app
+    container_name: try-cb-api
+  
+  frontend:
+      build: "https://github.com/osfameron/try-cb-frontend-v2.git#DOC-8135-travel-sample"
```
Comment (osfameron) [inline docker-compose.yml]: this is now on https://github.com/couchbaselabs/try-cb-frontend-v2

File: docker-compose.yml [hunk 2]
```diff
@@ -8,17 +8,17 @@ services:
...
+      - CB_USER
+      - CB_PSWD
+    command: $CB_HOST $CB_USER $CB_PSWD
     volumes:
       - .:/app
     container_name: try-cb-api
   
   frontend:
-      build: "https://github.com/osfameron/try-cb-frontend-v2.git#DOC-8135-travel-sample"
+      build: "https://github.com/couchbaselabs/try-cb-frontend-v2.git#7.0"
```
Comment (osfameron) [inline docker-compose.yml:15]: ah cool, it's done. Could be merged for clarity, but no biggie.

File: src/main/java/trycb/service/Airport.java [hunk 3]
```diff
@@ -1,32 +1,33 @@
...
     private static final Logger LOGGER = LoggerFactory.getLogger(Airport.class);
 
-
     /**
      * Find all airports.
      */
     public static Result<List<Map<String, Object>>> findAll(final Cluster cluster, final String bucket, String params) {
         StringBuilder builder = new StringBuilder();
-        builder.append("select airportname from `").append(bucket).append("` where ");
+        builder.append("SELECT airportname FROM `").append(bucket).append("`.inventory.airport WHERE ");
```
Comment (osfameron) [inline src/main/java/trycb/service/Airport.java:30]: that's depressing syntax isn't it, does Java still not have string templates?
I wonder if would be clearer with the appends on newlines, but is fine.

File: src/main/java/trycb/service/FlightPath.java [hunk 4]
```diff
@@ -26,22 +28,22 @@
...
-        builder.append("select faa as toAirport ");
-        builder.append("from `").append(bucket).append("` ");
-        builder.append("where airportname = $to");
+        builder.append("SELECT faa as fromAirport ");
+        builder.append("FROM `").append(bucket).append("`.inventory.airport ");
+        builder.append("WHERE airportname = $from ");
+        builder.append("UNION ");
+        builder.append("SELECT faa as toAirport ");
+        builder.append("FROM `").append(bucket).append("`.inventory.airport ");
+        builder.append("WHERE airportname = $to");
```
Comment (osfameron) [inline src/main/java/trycb/service/FlightPath.java:40]: :sob:

File: src/main/java/trycb/service/Hotel.java [hunk 5]
```diff
@@ -42,43 +43,27 @@ public Hotel(Bucket bucket) {
...
         if (location != null && !location.isEmpty() && !"*".equals(location)) {
-            fts.and(SearchQuery.disjuncts(
-                        SearchQuery.matchPhrase(location).field("country"),
-                        SearchQuery.matchPhrase(location).field("city"),
-                        SearchQuery.matchPhrase(location).field("state"),
-                        SearchQuery.matchPhrase(location).field("address")
-                ));
+            fts.and(SearchQuery.disjuncts(SearchQuery.matchPhrase(location).field("country"),
+                    SearchQuery.matchPhrase(location).field("city"), SearchQuery.matchPhrase(location).field("state"),
+                    SearchQuery.matchPhrase(location).field("address")));
```
Comment (osfameron) [inline src/main/java/trycb/service/Hotel.java]: previous formatting better!

File: src/main/java/trycb/service/Hotel.java [hunk 6]
```diff
@@ -42,43 +43,27 @@ public Hotel(Bucket bucket) {
...
         }
 
         if (description != null && !description.isEmpty() && !"*".equals(description)) {
-            fts.and(
-                SearchQuery.disjuncts(
-                        SearchQuery.matchPhrase(description).field("description"),
-                        SearchQuery.matchPhrase(description).field("name")
-                ));
+            fts.and(SearchQuery.disjuncts(SearchQuery.matchPhrase(description).field("description"),
+                    SearchQuery.matchPhrase(description).field("name")));
```
Comment (osfameron) [inline src/main/java/trycb/service/Hotel.java]: again. (being a clear example of FTS means this is important)

File: src/main/java/trycb/service/Hotel.java [hunk 7]
```diff
@@ -42,43 +43,27 @@ public Hotel(Bucket bucket) {
...
-        String ftsContext;
-        try {
-            ftsContext = JacksonTransformers.MAPPER.writerWithDefaultPrettyPrinter().
-                    writeValueAsString(fts.export());
-        } catch (JsonProcessingException e) {
-            ftsContext = fts.export().toString();
-        }
-        String subdocContext = "        Optional<LookupInResult> lookup = bucket.defaultCollection().lookupIn(row.id(),\n" +
-                "                Arrays.asList(get(\"country\"), get(\"city\"), get(\"state\"), get(\"address\"),\n" +
-                "                        get(\"name\"), get(\"description\")));";
```
Comment (osfameron) [inline src/main/java/trycb/service/Hotel.java:79]: don't quite understand this, but looks like you've got rid of some useless, ugly code, so yay!

File: src/main/java/trycb/service/Hotel.java [hunk 8]
```diff
@@ -109,9 +94,10 @@ public Hotel(Bucket bucket) {
 
             LookupInResult res;
             try {
-                res = bucket.defaultCollection().lookupIn(row.id(),
-                    Arrays.asList(get("country"), get("city"), get("state"), get("address"),
-                        get("name"), get("description")));
+                Scope scope = bucket.scope("inventory");
+                Collection collection = scope.collection("hotel");
+                res = collection.lookupIn(row.id(), Arrays.asList(get("country"), get("city"), get("state"),
+                        get("address"), get("name"), get("description")));
```
Comment (osfameron) [inline src/main/java/trycb/service/Hotel.java:106]: would be clearer with `Arrays` and `get()`on new lines

File: src/main/java/trycb/config/Request.java [hunk 9]
```diff
@@ -0,0 +1,19 @@
+package trycb.config;
+
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.web.filter.CommonsRequestLoggingFilter;
+
+@Configuration
+public class Request {
+
+    public @Bean CommonsRequestLoggingFilter logFilter() {
```
Comment (osfameron) [inline src/main/java/trycb/config/Request.java:14]: is this part related to User endpoint?

File: src/main/java/trycb/service/TenantUser.java [hunk 10]
```diff
@@ -21,80 +23,85 @@
...
-            throw new AuthenticationCredentialsNotFoundException("Bad Username or Password: " + username);
+            throw new AuthenticationCredentialsNotFoundException("Bad Username or Password");
         }
         JsonObject res = doc.contentAsObject();
-        if(BCrypt.checkpw(password, res.getString("password"))) {
-            return JsonObject.create()
-                .put("token", jwtService.buildToken(username))
-                .toMap();
+        if (BCrypt.checkpw(password, res.getString("password"))) {
+            Map<String, Object> data = JsonObject.create().put("token", jwtService.buildToken(username)).toMap();
```
Comment (osfameron) [inline src/main/java/trycb/service/TenantUser.java]: formatting with .put and .toMap clearer previously

File: src/main/java/trycb/service/TenantUser.java [hunk 11]
```diff
@@ -21,80 +23,85 @@
...
-    public Result<Map<String, Object>> createLogin(final Bucket bucket, final String username, final String password,
-            DurabilityLevel expiry) {
+    public Result<Map<String, Object>> createLogin(final Bucket bucket, final String tenant, final String username,
+            final String password, DurabilityLevel expiry) {
         String passHash = BCrypt.hashpw(password, BCrypt.gensalt());
-        JsonObject doc = JsonObject.create()
-            .put("type", "user")
-            .put("name", username)
-            .put("password", passHash);
+        JsonObject doc = JsonObject.create().put("type", "user").put("name", username).put("password", passHash);
```
Comment (osfameron) [inline src/main/java/trycb/service/TenantUser.java]: formatting

File: src/main/java/trycb/util/CorsFilter.java [hunk 12]
```diff
@@ -21,32 +21,41 @@
...
+        response.setHeader("Access-Control-Allow-Headers",
+                "Origin, X-Requested-With, Content-Type, Accept, Authorization");
+
         chain.doFilter(req, res);
     }
 
     @Override
-    public void init(FilterConfig filterConfig) throws ServletException {}
+    public void init(FilterConfig filterConfig) throws ServletException {
+    }
```
Comment (osfameron) [inline src/main/java/trycb/util/CorsFilter.java]: ok, though arguably `{}` suggests NOOP more clearly?

File: src/main/java/trycb/service/TenantUser.java [hunk 13]
```diff
@@ -21,80 +23,85 @@
...
         InsertOptions options = insertOptions();
         if (expiry.ordinal() > 0) {
             options.durability(expiry);
         }
-        String narration = "User account created in document " + username + " in bucket " + bucket.name()
-                + (expiry.ordinal() > 0 ? ", with expiry of " + expiry.ordinal() + "s" : "");
```
Comment (osfameron) [inline src/main/java/trycb/service/TenantUser.java:77]: this narration has gone?

File: src/main/java/trycb/web/AirportController.java [hunk 14]
```diff
@@ -33,8 +30,7 @@ public AirportController(Cluster cluster, Bucket bucket) {
     @RequestMapping
     public ResponseEntity<? extends IValue> airports(@RequestParam("search") String search) {
         try {
-            Result<List<Map<String, Object>>> result = Airport.findAll(cluster, bucket.name(), search);
-            return ResponseEntity.ok(result);
+            return ResponseEntity.ok(Airport.findAll(cluster, bucket.name(), search));
```
Comment (osfameron) [inline src/main/java/trycb/web/AirportController.java:37]: fine in one statement, perhaps `Airport.findAll` on newline?

File: src/main/java/trycb/web/TenantUserController.java [hunk 15]
```diff
@@ -88,7 +88,7 @@ public TenantUserController(Bucket bucket, TenantUser tenantUserService, TokenSe
             jwtService.verifyAuthenticationHeader(authentication, username);
             Result<Map<String, Object>> result = tenantUserService.registerFlightForUser(bucket, tenant, username,
                     jsonData.getArray("flights"));
-            return ResponseEntity.accepted().body(result);
+            return ResponseEntity.ok().body(result);
```
Comment (osfameron) [inline src/main/java/trycb/web/TenantUserController.java:99]: is this related to Swagger?

File: src/main/resources/application.properties [hunk 16]
```diff
@@ -7,3 +7,5 @@ storage.password=password
 #in seconds, set to 0 to disable
 storage.expiry=0
 logging.level.org.springframework.web.filter.CommonsRequestLoggingFilter=DEBUG
+springdoc.swagger-ui.url=/swagger.json
+springdoc.swagger-ui.path=/apidocs
```
Comment (osfameron) [inline src/main/resources/application.properties:11]: wow, so just these 2 config files and it all JustWorks?

File: src/main/java/trycb/web/TenantUserController.java [hunk 17]
```diff
@@ -108,9 +118,11 @@ public Object booked(@PathVariable("tenant") String tenant, @PathVariable("usern
             jwtService.verifyAuthenticationHeader(authentication, username);
             return ResponseEntity.ok(tenantUserService.getFlightsForUser(bucket, tenant, username));
         } catch (IllegalStateException e) {
+            LOGGER.error("Failed with invalid state exception", e);
             return ResponseEntity.status(HttpStatus.FORBIDDEN)
                     .body(new Error("Forbidden, you don't have access to this cart"));
         } catch (IllegalArgumentException e) {
+            LOGGER.error("Failed with invalid argument exception", e);
```
Comment (osfameron) [inline src/main/java/trycb/web/TenantUserController.java:125]: minor q: why is an IllegalArgumentException thrown to indicate this? (not important, just curious)

**Comment by osfameron** [comment]
This looks great! 👍

**Review by daschl — APPROVED** [review]
this is looking very solid @maria-robobug  good job :) .. I think we can merge this if you think it's ready to go and then just iterate on smaller changes in the branch?

File: src/main/java/trycb/service/TenantUser.java (hunk 13 above)
Comment (maria-robobug) [reply src/main/java/trycb/service/TenantUser.java:77]: I updated this to match the narration we use for Python and Nodejs to keep the output we send to the FE the same:

See line [81](https://github.com/couchbaselabs/try-cb-java/pull/16/files/18b76fd03c4f5e8f09d0c4a03b1c952a92546105..114920fda6d26e41ebea1af3324c32c2e027f5cd#diff-982bc6359b182c2a5bb7995bf0fa11e9b6f31a4ead60046b44a2730e55fa4d2fR81)

File: src/main/java/trycb/web/TenantUserController.java (hunk 15 above)
Comment (maria-robobug) [reply src/main/java/trycb/web/TenantUserController.java:99]: Yeah, we return 200 in the swagger definition, so just updating to make sure we're doing the same here.

File: src/main/resources/application.properties (hunk 16 above)
Comment (maria-robobug) [reply src/main/resources/application.properties:11]: yeah the application properties from the swagger library allow us to override the default url paths quite easily.

File: src/main/java/trycb/web/TenantUserController.java (hunk 17 above)
Comment (maria-robobug) [reply src/main/java/trycb/web/TenantUserController.java:125]: Probably was there as good practice to check if the user has passed an invalid argument to the endpoint.

File: src/main/java/trycb/config/Request.java (hunk 9 above)
Comment (maria-robobug) [reply src/main/java/trycb/config/Request.java:14]: It's just so we can see requests made to the backend in the logs, was quite hard to debug issues without.

File: src/main/java/trycb/service/Hotel.java (hunk 7 above)
Comment (maria-robobug) [reply src/main/java/trycb/service/Hotel.java:79]: I think it was trying to show the code that gets executed in the narration via a _string_ representation, but it just looks weird and confusing.

Plus a user can just see the code in the app, and it would be a pain to update this if the code changes.

This was what it was trying to show I think (but see already the code changed and the string hadn't been updated): https://github.com/couchbaselabs/try-cb-java/blob/cc831f4d9609d1dd3bd8e62ff9b819e139e788db/src/main/java/trycb/service/Hotel.java#L110-L114

Comment (osfameron) [reply src/main/java/trycb/service/Hotel.java:79]: Wow, yes, that seems ill-advised, I wonder if there was a specific request at the time to focus on that feature... anyway, good call to remove it :D
